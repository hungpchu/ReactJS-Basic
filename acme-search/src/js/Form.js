import React from "react";
import ".././styles/Form.css";

class Form extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      dataSourceValue: "select",
      dropdownValue: ["select", "contacts", "dropbox", "slack", "tweet"],
      contactForm: [
        "id",
        "name",
        "company",
        "emails",
        "phones",
        "matching_terms",
        "last_contact",
      ],
      slackForm: [
        "id",
        "channel",
        "alice",
        "message",
        "timestamp",
        "matching_terms",
      ],
      tweetForm: ["user", "message", "timestamp", "matching_terms"],
      dropboxForm: [
        "id",
        "path",
        "title",
        "shared_with",
        "matching_terms",
        "created",
      ],
      array: ["emails", "phones", "shared_with", "matching_terms"],
      newObject: {},
    };
    this.handleSelectionChange = this.handleSelectionChange.bind(this);
    this.displayForm = this.displayForm.bind(this);
    this.updateNewObject = this.updateNewObject.bind(this);
    this.saveData = this.saveData.bind(this);
    this.displayDropDownValue = this.displayDropDownValue.bind(this);
  }

  handleSelectionChange(event) {
    this.setState({ dataSourceValue: event.target.value });
    this.setState({ newObject: {} });
  }

  updateNewObject(key, event) {
    const isArray = this.state.array.includes(key.element);
    const updatedNewObject = this.state.newObject;
    isArray
      ? (updatedNewObject[key.element] = event.target.value.split(","))
      : (updatedNewObject[key.element] = event.target.value);
    this.setState({ newObject: updatedNewObject });
  }

  displayForm() {
    if (this.state.dataSourceValue === "select") return null;
    const array =
      this.state.dataSourceValue === "contacts"
        ? this.state.contactForm
        : this.state.dataSourceValue === "dropbox"
        ? this.state.dropboxForm
        : this.state.dataSourceValue === "slack"
        ? this.state.slackForm
        : this.state.tweetForm;
    const field = array.map((element) => (
      <li key={JSON.stringify(element)} id={element + "-field-item"}>
        <label id={element + "-label"}>{element}</label>
        <input
          type="text"
          id={element + "-input"}
          placeholder={element}
          onChange={this.updateNewObject.bind(this, { element })}
        />
        {this.state.array.includes(element) && (
          <p className="instruction-message">
            *{element} is array. Please add ',' in between for multiple value
          </p>
        )}
      </li>
    ));
    return (
      <>
        <ul className="field-input-list">{field}</ul>
        <button id="save-data-button" onClick={this.saveData}>
          Save data
        </button>
      </>
    );
  }

  saveData() {
    this.props.saveUpdatedData(
      this.state.dataSourceValue,
      this.state.newObject
    );
    // set the newObject back to empty
    // for the new updated one later
    this.setState({ newObject: {} });
  }

  displayDropDownValue() {
    return this.state.dropdownValue.map((value) => (
      <option
        id={value + "-selection-value"}
        type="submit"
        key={value + "-selection-value"}
      >
        {value}
      </option>
    ));
  }

  render() {
    return (
      <div className="update-form-body">
        <label className="dropdown-label">
          Choose a datasource to save updated data:
        </label>
        <select
          className="dropdown-selection"
          onClick={this.handleSelectionChange}
        >
          {this.displayDropDownValue()}
        </select>
        {this.displayForm()}
      </div>
    );
  }
}

export default Form;
