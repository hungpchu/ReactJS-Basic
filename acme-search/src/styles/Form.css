select {
  font-size: 0.9rem;
  padding: 2px 5px;
}

ul > li {
  list-style-type: none;
}

/* for the label in form */
li > label {
  display: inline-block;
  color: white;
  width: 40px;
  margin-bottom: 20px;
  margin-left: -40px;
}

.instruction-message {
  color: red;
  font-size: 0.8rem;
  margin-top: -15px;
  margin-left: 100px;
}

.update-form-body {
  /* Center the form on the page */
  width: 600px;
  /* Form outline */
  padding: 1em;
  border: 1px solid white;
  border-radius: 1em;
  margin: 0 auto;
}

.update-form-body > label {
  color: whitesmoke;
  font-size: 1.2rem;
  padding-right: 10px;
}

#save-data-button {
  margin-left: 500px;
}
