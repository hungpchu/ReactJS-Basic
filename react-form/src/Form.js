import React from 'react';


class Form extends React.Component{
    constructor(props)
    {
        super(props);
        this.state = 
        { 
            text: "",
            age:  "" 
        };
    }

   handleChangeName = (event) => {
       this.setState({text : event.target.value});
   }

   handleChangeAge = (event) => {
    this.setState({age: event.target.value});
}

   handleSubmit = ( event ) => {
       alert("Value of submission text : " + this.state.text + " Value of submission age: " + this.state.age);
       event.preventDefault();
   }

   render() {

    return(
        
        <form onSubmit={this.handleSubmit}>
            <label>
                Name: <input type="text" value={this.state.text} onChange={this.handleChangeName} />
                </label>
                <br/>
                <label>
                Age: <input type="text" value={this.state.age} onChange={this.handleChangeAge} />
            </label>
            <br/>
            <input type="submit" value="Submit" />
        </form>
    );
   }
}

export default Form;