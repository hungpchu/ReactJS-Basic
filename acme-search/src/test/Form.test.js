import React from "react";
import renderer from "react-test-renderer";
import Form from "./../js/Form.js";
import Adapter from "enzyme-adapter-react-16";
import { mount, configure } from "enzyme";

configure({ adapter: new Adapter() });
describe("<Form />", () => {
  let mockForm;
  beforeEach(() => {
    mockForm = mount(<Form />);
  });
  describe("(snapshot)", () => {
    it("renders correctly", () => {
      const mockSaveUpdatedData = jest.fn();
      const formSnapShot = renderer
        .create(<Form saveUpdatedData={mockSaveUpdatedData} />)
        .toJSON();
      expect(formSnapShot).toMatchSnapshot();
    });
  });

  describe("the visibility of the field-input-list", () => {
    test("should not render field-input-list if value of dataSourceValue is select", () => {
      mockForm.setState({
        dataSourceValue: "select",
      });
      expect(mockForm.find(".field-input-list").exists()).toEqual(false);
    });

    test("should render field-input-list if value of dataSourceValue is not select", () => {
      mockForm.setState({
        dataSourceValue: "contacts",
      });
      expect(mockForm.find(".field-input-list").exists()).toEqual(true);
    });
  });

  describe("the function saveUpdatedData from Search", () => {
    test("should be called if value of dataSourceValue is not select", () => {
      let mockSaveUpdatedData = jest.fn();
      mockForm.setProps({
        saveUpdatedData: mockSaveUpdatedData,
      });
      mockForm.setState({
        dataSourceValue: "contacts",
      });
      mockForm.find("#save-data-button").simulate("click");
      expect(mockSaveUpdatedData).toHaveBeenCalled();
    });
  });
});
