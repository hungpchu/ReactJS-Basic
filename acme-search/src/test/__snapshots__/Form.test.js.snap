// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> (snapshot) renders correctly 1`] = `
<div
  className="update-form-body"
>
  <label
    className="dropdown-label"
  >
    Choose a datasource to save updated data:
  </label>
  <select
    className="dropdown-selection"
    onClick={[Function]}
  >
    <option
      id="select-selection-value"
      type="submit"
    >
      select
    </option>
    <option
      id="contacts-selection-value"
      type="submit"
    >
      contacts
    </option>
    <option
      id="dropbox-selection-value"
      type="submit"
    >
      dropbox
    </option>
    <option
      id="slack-selection-value"
      type="submit"
    >
      slack
    </option>
    <option
      id="tweet-selection-value"
      type="submit"
    >
      tweet
    </option>
  </select>
</div>
`;
